File: mechLIB/dcm_t.cs
using System.Numerics;

namespace mechLIB
{
    public class dcm_t
    {
        public float Xx;
        public float Yx;
        public float Zx;

        public float Xy;
        public float Yy;
        public float Zy;

        public float Xz;
        public float Yz;
        public float Zz;
        public dcm_t(Vector3 endPoint, Vector3 radiusPoint)
        {
            Xx = endPoint.X / endPoint.Length();
            Yx = endPoint.Y / endPoint.Length();
            Zx = endPoint.Z / endPoint.Length();

            Xy = radiusPoint.X / radiusPoint.Length();
            Yy = radiusPoint.Y / radiusPoint.Length();
            Zy = radiusPoint.Z / radiusPoint.Length();

            Xz = Yx * Zy - Zx * Yy;
            Yz = 0 - (Xx * Zy - Zx * Xy);
            Zz = Xx * Yy - Yx * Xy;
        }

        public void ToGlob(Vector3 Lp, ref Vector3 gA)
        {
            gA.X = Xx * Lp.X + Xy * Lp.Y + Xz * Lp.Z;
            gA.Y = Yx * Lp.X + Yy * Lp.Y + Yz * Lp.Z;
            gA.Z = Zx * Lp.X + Zy * Lp.Y + Zz * Lp.Z;
        }

        public void ToLoc(Vector3 Gp, ref Vector3 lA)
        {
            lA.X = Xx * Gp.X + Yx * Gp.Y + Zx * Gp.Z;
            lA.Y = Xy * Gp.X + Yy * Gp.Y + Zy * Gp.Z;
            lA.Z = Xz * Gp.X + Yz * Gp.Y + Zz * Gp.Z;
        }
    }
}

File: mf/dcm_t.cpp
#include "dcm_t.h"
#include "coords.h"

dcm_t::dcm_t(xyz_t endPoint, xyz_t radiusPoint)
{
	float lax = crds::GetTotL(endPoint);
	float lby = crds::GetTotL(radiusPoint);
	Xx = endPoint.x / lax;
	Yx = endPoint.y / lax;
	Zx = endPoint.z / lax;
	Xy = radiusPoint.x / lby;
	Yy = radiusPoint.y / lby;
	Zy = radiusPoint.z / lby;
	Xz = Yx * Zy - Zx * Yy;
	Yz = 0 - (Xx * Zy - Zx * Xy);
	Zz = Xx * Yy - Yx * Xy;
}

inline void dcm_t::ToGlob(xyz_t Lp, xyz_t& gA)
{
	gA.x = Xx * Lp.x + Xy * Lp.y + Xz * Lp.z;
	gA.y = Yx * Lp.x + Yy * Lp.y + Yz * Lp.z;
	gA.z = Zx * Lp.x + Zy * Lp.y + Zz * Lp.z;
}

inline void dcm_t::ToLoc(xyz_t Gp, xyz_t& lA)
{
	lA.x = Xx * Gp.x + Yx * Gp.y + Zx * Gp.z;
	lA.y = Xy * Gp.x + Yy * Gp.y + Zy * Gp.z;
	lA.z = Xz * Gp.x + Yz * Gp.y + Zz * Gp.z;
}
